### PR TITLE
arrow-cpp: 13.0.0 -> 14.0.1

### LINKED ISSUE #266320: Update request: arrow-cpp 13.0.0 → 14.0.1 (CVE-2023-47248)
https://www.openwall.com/lists/oss-security/2023/11/08/7
Arbitrary code execution in PyArrow (CVE-2023-47248)

- Package name: python311Packages.pyarrow (src comes from arrow-cpp)
- Latest released version: 14.0.1

- Current version on the unstable channel: 13.0.0
- Current version on the stable/release channel: 12.0.0

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**

@tobim 
@veprbl 
@cpcloud 

-----

Note for maintainers: Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by tobim**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

resolves #266320

Targetting staging-next until that branch is merged.

### CODE REVIEW

**Comment by risicle** [comment]
Struggling to see the security relevance of this release..

**Comment by cpcloud** [comment]
@risicle https://nvd.nist.gov/vuln/detail/CVE-2023-47248

**Comment by risicle** [comment]
Thanks. Hopefully we can fix stable without resorting to that "hotfix" monkeypatch ;)

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 1]
```diff
@@ -115,15 +115,10 @@ stdenv.mkDerivation rec {
...
     repo = "substrait";
-    rev = "v0.20.0";
-    hash = "sha256-71hAwJ0cGvpwK/ibeeQt82e9uqxcu9sM1rPtPENMPfs=";
+    rev = "v0.27.0";
+    hash = "sha256-wptEAXembah04pzqAz6UHeUxp+jMf6Lh/IdyuIhy/a8=";
   };
 
-  patches = [
-    # Protobuf switched to lower case project name.
-    ./cmake-find-protobuf.patch
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix:124]: Need to remove the file.

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.pyarrow

**Comment by risicle** [comment]
`nixpkgs-review` reveals no new failures, macos 10.15

**Comment by risicle** [comment]
Any chance of having this come out of draft?

I would suggest applying the patch to master otherwise, but it would be nice to get v14 into 23.11.

**Comment by tobim** [comment]
I hadn't noticed that staging-next got merged yet. I'll quickly remove the unused patch as @veprbl pointed out.

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.pyarrow

**Review by veprbl — APPROVED** [review]

**Comment by risicle** [comment]
@ofborg build python3Packages.streamz

**Review by risicle — APPROVED** [review]
`nixpkgs-review` reveals no new failures, nixos x86_64

**Pull request merged**